### PR TITLE
Emoji Data Restructuring

### PR DESCRIPTION
**Pull request by KennethMetz**
Reworks where emoji data is stored to allow for sorting of comments/reviews by popularity.
- Whether a user has selected an emoji is stored within `users/:uid/reactions/:uniqueEntityId`
- The emoji tally for an item is stored within either `animeData` or `watchlistData` collection, in the same doc as the review/comment information.

Quirks to be aware of:
- On clicking an emoji, we only write to firestore. The visual update of the reaction count on the chip is faked, to eliminate a delay between clicking the chip and seeing the value increment.  Obviously this has serious flaws (ie. the firestore write fails). Will revisit.
- To update the emoji tally I'm using firestore's increment function which can only be called [once per second](https://firebase.blog/posts/2019/03/increment-server-side-cloud-firestore/). I will update to using a [distributed counter](https://firebase.google.com/docs/firestore/solutions/counters) on a future PR (this one's bursting at the seams already!)

Next steps for Ken after we merge:
- Write firestore rules.
- Move data for existing lists/comments/reviews to match new data structure.

### CODE REVIEW

**Review by metzsg — COMMENTED** [review]
I really like this update -- moves the data storage structure to something much more scalable and locked-down.  Nice!  It is a big change and I have been working through reading it.  Not quite done yet, but I've left some first-pass comments here for you to check out.  Lmk what you think!

File: src/Components/Firestore.js
Comment (metzsg) [reply src/Components/Firestore.js:211]: Then a usage in Review.js could be 
```js
  useEffect(() => {
    getReviewReactionsNew(user.uid, uniqueEntityId)
      .then((value) => setItem(value))
      .catch(() => {
        // Show error UI, maybe.
      });
  }, [user.uid, uniqueEntityId]);
```

Comment (KennethMetz) [reply src/Components/Firestore.js:211]: Yeah, that's a fair point - it does seem to make things messier. I'll revise accordingly!

File: src/Components/Review.js
Comment (KennethMetz) [reply src/Components/Review.js:46]: Lol, agreed! This definitely tripped me up during my re-work of it.  

Yes - it should only store the booleans for the users reactions.  I'll definitely incorporate this - future Ken will thank me, I'm sure!

File: src/Components/ProfileListPage.js
Comment (KennethMetz) [reply src/Components/ProfileListPage.js:75]: Kind of - I haven't moved the firestore data for any of the existing watchlists/reviews yet, so this behavior is expected for watchlists/reviews created prior to this PR.

I was waiting until we were ready to merge to edit any existing data.

For a demonstration of a working watchlist checkout out @TwitterTestAccount's list 'Shonen Jump FTW'.

File: src/Components/Firestore.js
Comment (KennethMetz) [reply src/Components/Firestore.js:254]: We'll assume this is an invariant going forward (all watchlists/reviews/comments will have that field created during their initial save).

The downside of this is what you're finding...all pre-existing watchlists/reviews/comments won't show valid counts until I update the data manually.

File: src/Components/Firestore.js
Comment (metzsg) [inline src/Components/Firestore.js:229]: Is this function the same as `getReviewReactions` above?  If so, do we need both, or should we consolidate to just one?  If so, let's use the generally-named `getUserReactions`

File: src/Components/Firestore.js
Comment (metzsg) [inline src/Components/Firestore.js:245]: This function only gets reaction counts for watchlists, not for reviews or comments -- should that be clear in the name?

File: src/Components/Firestore.js
Comment (metzsg) [inline src/Components/Firestore.js:263]: How about adding:
```js
else throw new Error(`Unknown reaction: ${reaction}`);
```
so we don't proceed if the reaction type is bad

File: src/Components/Firestore.js
Comment (metzsg) [inline src/Components/Firestore.js:277]: Similarly, should we throw if `reactionTo` is not known, since we would be proceeding with an undefined `rxnCountRef`, which will surely cause problems?

File: src/Components/Firestore.js
Comment (metzsg) [inline src/Components/Firestore.js:303]: This will require one user (the user doing the deletion) to be able to delete documents from a bunch of other users' `/users/:uid/reactions` collections.  I know we discussed an access-control principal that the owning user is the only one able to write to their `/users/:uid` document and subcollections, and so these deletions would be disallowed if we add such a rule.  

I'm also worried this could result in a lot of deletions all happening from one user's web app, in the case of a review that had a lot of reactions being deleted.  Is there any harm in leaving these zombie reactions in the respective users' `reactions` collections?  Maybe at least util a server-side vacuum job can clean them up?

**Review by metzsg — APPROVED** [review]
Thanks for the updates, this is really shaping up nicely!  Left a few final comments, ptal when you have a sec

File: src/Components/Firestore.js
Comment (KennethMetz) [reply src/Components/Firestore.js:229]: Great catch!! Functions have been consolidated.

File: src/Components/Firestore.js
Comment (KennethMetz) [reply src/Components/Firestore.js:303]: All very fair concerns!

The unwanted behavior from zombie reactions, is if someone updates a popular review/comment by deleting it and then reposting (instead of editing it - the way god intended) then all of their friends who had interacted with the old review/comment will still have those emojis marked as selected, despite the emoji count of this new review/comment being 0. 

Admittedly, this is probably an unlikely case, but I did run into it while testing things out so take that for what you will 🙂.  I'm satisfied it we just run a server-side sweeper daily or so.. 

Lmkwyt...but for now I'll delete this `deleteAllReactions` function.

**Pull request merged**
